teatest/teatest.go

```diff
@@ -0,0 +1,123 @@
+// Package teatest provides helper functions to test tea.Model's.
+package teatest
+
+import (
+	"bytes"
+	"flag"
+	"io"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"sync"
+	"syscall"
+	"testing"
+
+	tea "github.com/charmbracelet/bubbletea"
+)
+
+// Program defines the subset of the tea.Program API we need for testing.
+type Program interface {
+	Send(tea.Msg)
+}
+
+// TestModel tests a given model with the given interactions and assertions.
+func TestModel(
+	tb testing.TB,
+	m tea.Model,
+	interact func(p Program, in io.Writer),
+	assert func(out []byte),
+) {
+	var in bytes.Buffer
+	var out bytes.Buffer
+
+	p := tea.NewProgram(m, tea.WithInput(&in), tea.WithOutput(safe(&out)), tea.WithoutSignals())
+
+	ints := make(chan os.Signal, 1)
+	signal.Notify(ints, syscall.SIGINT)
+	done := make(chan bool, 1)
+	go func() {
+		if err := p.Start(); err != nil {
+			tb.Fatalf("app failed: %s", err)
+		}
+		done <- true
+	}()
+	go func() {
+		<-ints
+		signal.Stop(ints)
+		tb.Log("interrupted")
+		p.Quit()
+	}()
+
+	// run the user interactions and then force a quit
+	interact(p, safe(&in))
+	p.Quit()
+	if err := p.ReleaseTerminal(); err != nil {
+		tb.Fatalf("could not restore terminal: %v", err)
+	}
+
+	// wait for the program to quit and assert
+	<-done
+	assert(out.Bytes())
+}
+
+// TypeText types the given text into the given program.
+func TypeText(p Program, s string) {
+	for _, c := range []byte(s) {
+		p.Send(tea.KeyMsg{
+			Runes: []rune{rune(c)},
+			Type:  tea.KeyRunes,
+		})
+	}
```

Comment:
It's would also be nice to be able to send `tea.KeyMsg`s of different types, e.g. for `tea.KeyCtrlC`, etc. There's an initial implementation at https://github.com/twpayne/chezmoi/blob/4f4760f4911ff03fe431ce31bd847a05b213d6a9/pkg/chezmoibubbles/chezmoibubbles_test.go. This converts a string like `"a\r"` to the equivalent of the user pressing `a` then `Enter`.

Reply:
That seems to make sense as well, will play with it later today 🙏